packages/multicast_dns/lib/multicast_dns.dart

```diff
@@ -0,0 +1,200 @@
+
+    if (_started || _starting) {
+      return;
+    }
+    _starting = true;
+
+    // Listen on all addresses.
+    _incoming = await RawDatagramSocket.bind(
+      listenAddress.address,
+      mDnsPort,
+      reuseAddress: true,
+      reusePort: true,
+      ttl: 255,
+    );
+
+    // Can't send to IPv6 any address.
+    if (_incoming.address != InternetAddress.anyIPv6) {
+      _sockets.add(_incoming);
+    }
+
+    _mDnsAddress = _incoming.address.type == InternetAddressType.IPv4
+        ? mDnsAddressIPv4
+        : mDnsAddressIPv6;
+
+    final List<NetworkInterface> interfaces =
+        await interfacesFactory(listenAddress.type);
+
+    for (NetworkInterface interface in interfaces) {
+      // Create a socket for sending on each adapter.
+      final InternetAddress targetAddress = interface.addresses[0];
+      final RawDatagramSocket socket = await RawDatagramSocket.bind(
+        targetAddress,
+        mDnsPort,
+        reuseAddress: true,
+        reusePort: true,
+        ttl: 255,
+      );
+      _sockets.add(socket);
+      // Ensure that we're using this address/interface for multicast.
+      if (targetAddress.type == InternetAddressType.IPv4) {
+        socket.setRawOption(RawSocketOption(
+          RawSocketOption.levelIPv4,
+          RawSocketOption.IPv4MulticastInterface,
+          targetAddress.rawAddress,
+        ));
+      } else {
+        socket.setRawOption(RawSocketOption.fromInt(
+          RawSocketOption.levelIPv6,
+          RawSocketOption.IPv6MulticastInterface,
+          interface.index,
+        ));
+      }
+      // Join multicast on this interface.
+      _incoming.joinMulticast(_mDnsAddress, interface);
+    }
+    _incoming.listen(_handleIncoming);
+    _started = true;
+    _starting = false;
+  }
+
+  /// Stop the client and close any associated sockets.
+  void stop() {
+    if (!_started) {
+      return;
+    }
+    if (_starting) {
+      throw StateError('Cannot stop mDNS client while it is starting.');
+    }
+
+    for (RawDatagramSocket socket in _sockets) {
+      socket.close();
+    }
+
+    _resolver.clearPendingRequests();
+
+    _started = false;
+  }
+
+  /// Lookup a [ResourceRecord], potentially from cache.
+  ///
+  /// The [type] parameter must be a valid [ResourceRecordType].  The [name] parameter is
+  /// the name of the service to lookup, and must not be null.  The [timeout]
+  /// parameter specifies how long the intenral cache should hold on to the
+  /// record.  The [multicast] parameter specifies whether the query should be
+  /// sent as unicast (QU) or multicast (QM).
+  ///
+  /// Note that some publishers have been observed to not respond to unicast
+  /// requests properly, so the default is true.
+  Stream<T> lookup<T extends ResourceRecord>(
+    ResourceRecordQuery query, {
+    Duration timeout = const Duration(seconds: 5),
+  }) {
+    if (!_started) {
+      throw StateError('mDNS client is not started.');
```

Comment:
Maybe 'mDNS client must be started before calling lookup'? Describes better what's wrong.